code/datums/elements/strippable.dm

```diff
@@ -0,0 +1,531 @@
+/datum/element/strippable
+	element_flags = ELEMENT_BESPOKE | ELEMENT_DETACH
+	id_arg_index = 2
+
+	/// An assoc list of keys to /datum/strippable_item
+	var/list/items
+
+	/// A proc path that returns TRUE/FALSE if we should show the strip panel for this entity.
+	/// If it does not exist, the strip menu will always show.
+	/// Will be called with (mob/user).
+	var/should_strip_proc_path
+
+	/// An existing strip menus
+	var/list/strip_menus
+
+/datum/element/strippable/Attach(datum/target, list/items, should_strip_proc_path)
+	. = ..()
+	if (!isatom(target))
+		return ELEMENT_INCOMPATIBLE
+
+	RegisterSignal(target, COMSIG_ATOM_DROP_ON, PROC_REF(mouse_drop_onto))
+
+	src.items = items
+	src.should_strip_proc_path = should_strip_proc_path
+
+/datum/element/strippable/Detach(datum/source, force)
+	. = ..()
+
+	UnregisterSignal(source, COMSIG_ATOM_DROP_ON)
+
+	if (!isnull(strip_menus))
+		QDEL_NULL(strip_menus[source])
+
+/datum/element/strippable/proc/mouse_drop_onto(datum/source, atom/over, mob/user)
+	SIGNAL_HANDLER
+	if (user == source)
+		return
+
+	if (over == source)
+		return
+
+	var/mob/overmob = over
+	if (!istype(overmob))
+		return
+
+	if (!overmob.Adjacent(source))
+		return
+
+	if (!overmob.client)
+		return
+
+	if (overmob.client != user)
+		return
+
+	if (!isnull(should_strip_proc_path) && !call(source, should_strip_proc_path)(overmob))
+		return
+
+	var/datum/strip_menu/strip_menu
+
+	if (isnull(strip_menu))
+		strip_menu = new(source, src)
+		LAZYSET(strip_menus, source, strip_menu)
+
+	INVOKE_ASYNC(strip_menu, PROC_REF(tgui_interact), overmob)
+
+/// A representation of an item that can be stripped down
+/datum/strippable_item
+	/// The STRIPPABLE_ITEM_* key
+	var/key
+
+	/// Should we warn about dangerous clothing?
+	var/warn_dangerous_clothing = TRUE
+
+/// Gets the item from the given source.
+/datum/strippable_item/proc/get_item(atom/source)
+
+/// Tries to equip the item onto the given source.
+/// Returns TRUE/FALSE depending on if it is allowed.
+/// This should be used for checking if an item CAN be equipped.
+/// It should not perform the equipping itself.
+/datum/strippable_item/proc/try_equip(atom/source, obj/item/equipping, mob/user)
+	if ((equipping.flags_item & ITEM_ABSTRACT))
+		return FALSE
+	if ((equipping.flags_item & NODROP))
+		to_chat(user, SPAN_WARNING("You can't put [equipping] on [source], it's stuck to your hand!"))
+		return FALSE
+	if (ishuman(source))
+		var/mob/living/carbon/human/sourcehuman = source
+		if(HAS_TRAIT(sourcehuman, TRAIT_UNSTRIPPABLE) && !sourcehuman.is_mob_incapacitated())
+			to_chat(src, SPAN_DANGER("[sourcehuman] is too strong to force [equipping] onto them!"))
+			return
+	return TRUE
+
+/// Start the equipping process. This is the proc you should yield in.
+/// Returns TRUE/FALSE depending on if it is allowed.
+/datum/strippable_item/proc/start_equip(atom/source, obj/item/equipping, mob/user)
```

Comment:
Equipping followed by then simultaneous stripping of something (as a rifleman) does not say "You can't do this right now." but it fails. It probably should just say that message rather than initiating a do_after.